File: src-tauri/src/yaml_config.rs
use crate::models::{Group, Project};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::{Path, PathBuf};

const YAML_FILENAME: &str = "openrunner.yaml";
const YML_FILENAME: &str = "openrunner.yml";

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct YamlConfig {
    pub version: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub env_vars: Option<HashMap<String, String>>,
    pub projects: Vec<YamlProject>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct YamlProject {
    pub name: String,
    pub command: String,
    #[serde(rename = "type")]
    pub project_type: YamlProjectType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auto_restart: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cwd: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub interactive: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub env_vars: Option<HashMap<String, String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum YamlProjectType {
    Service,
    Task,
}

impl From<YamlProjectType> for crate::models::ProjectType {
    fn from(yaml_type: YamlProjectType) -> Self {
        match yaml_type {
            YamlProjectType::Service => crate::models::ProjectType::Service,
            YamlProjectType::Task => crate::models::ProjectType::Task,
        }
    }
}

impl From<crate::models::ProjectType> for YamlProjectType {
    fn from(project_type: crate::models::ProjectType) -> Self {
        match project_type {
            crate::models::ProjectType::Service => YamlProjectType::Service,
            crate::models::ProjectType::Task => YamlProjectType::Task,
        }
    }
}

pub fn find_yaml_file(directory: &Path) -> Option<PathBuf> {
    let yaml_path = directory.join(YAML_FILENAME);
    if yaml_path.exists() {
        return Some(yaml_path);
    }

    let yml_path = directory.join(YML_FILENAME);
    if yml_path.exists() {
        return Some(yml_path);
    }

    None
}

pub fn get_yaml_path(directory: &Path) -> PathBuf {
    let yaml_path = directory.join(YAML_FILENAME);
    if yaml_path.exists() {
        return yaml_path;
    }
    directory.join(YML_FILENAME)
}

pub fn parse_yaml(path: &Path) -> Result<YamlConfig, String> {
    let content =
        std::fs::read_to_string(path).map_err(|e| format!("Failed to read YAML file: {}", e))?;

    let config: YamlConfig =
        serde_yaml::from_str(&content).map_err(|e| format!("Failed to parse YAML file: {}", e))?;

    Ok(config)
}

pub fn write_yaml(group: &Group, path: &Path) -> Result<(), String> {
    let yaml_projects: Vec<YamlProject> = group
        .projects
        .iter()
        .map(|p| YamlProject {
            name: p.name.clone(),
            command: p.command.clone(),
            project_type: p.project_type.into(),
            auto_restart: if p.auto_restart { Some(true) } else { None },
            cwd: p.cwd.clone(),
            interactive: if p.interactive { Some(true) } else { None },
            env_vars: if p.env_vars.is_empty() {
                None
            } else {
                Some(p.env_vars.clone())
            },
        })
        .collect();

    let yaml_config = YamlConfig {
        version: "1.0".to_string(),
        name: group.name.clone(),
        env_vars: if group.env_vars.is_empty() {
            None
        } else {
            Some(group.env_vars.clone())
        },
        projects: yaml_projects,
    };

    let yaml_content = serde_yaml::to_string(&yaml_config)
        .map_err(|e| format!("Failed to serialize YAML: {}", e))?;

    std::fs::write(path, yaml_content).map_err(|e| format!("Failed to write YAML file: {}", e))?;

    Ok(())
}

pub fn yaml_to_group(
    config: &YamlConfig,
    directory: &std::path::Path,
    sync_file: &std::path::Path,
) -> Group {
    let directory_str = directory.to_string_lossy().to_string();
    let sync_file_str = sync_file.to_string_lossy().to_string();
    group_from_yaml(config.clone(), directory_str, sync_file_str)
}

pub fn yaml_to_project(
    yaml_project: &YamlProject,
    _base_dir: &std::path::Path,
) -> crate::models::Project {
    crate::models::Project {
        id: uuid::Uuid::new_v4().to_string(),
        name: yaml_project.name.clone(),
        command: yaml_project.command.clone(),
        project_type: yaml_project.project_type.clone().into(),
        auto_restart: yaml_project.auto_restart.unwrap_or(true),
        cwd: yaml_project.cwd.clone(),
        interactive: yaml_project.interactive.unwrap_or(false),
        env_vars: yaml_project.env_vars.clone().unwrap_or_default(),
    }
}

pub fn group_from_yaml(config: YamlConfig, directory: String, sync_file: String) -> Group {
    let projects: Vec<Project> = config
        .projects
        .into_iter()
        .map(|p| Project {
            id: uuid::Uuid::new_v4().to_string(),
            name: p.name,
            command: p.command,
            project_type: p.project_type.into(),
            auto_restart: p.auto_restart.unwrap_or(true),
            cwd: p.cwd,
            interactive: p.interactive.unwrap_or(false),
            env_vars: p.env_vars.unwrap_or_default(),
        })
        .collect();

    Group {
        id: uuid::Uuid::new_v4().to_string(),
        name: config.name,
        directory,
        projects,
        env_vars: config.env_vars.unwrap_or_default(),
        sync_file: Some(sync_file),
        sync_enabled: true,
    }
}

pub fn sync_yaml(group: &Group, state: &crate::state::AppState) -> Result<(), String> {
    if !group.sync_enabled {
        return Ok(());
    }

    if let Some(ref yaml_path) = group.sync_file {
        let path = Path::new(yaml_path);
        write_yaml(group, path)?;

        if let Ok(watcher) = state.yaml_watcher.lock() {
            let _ = watcher.update_yaml_timestamp(yaml_path);
        }
    }
    Ok(())
}

pub fn add_project_to_yaml(
    group: &Group,
    _project: &Project,
    state: &crate::state::AppState,
) -> Result<(), String> {
    if !group.sync_enabled {
        return Ok(());
    }

    if let Some(ref yaml_path) = group.sync_file {
        let path = Path::new(yaml_path);
        write_yaml(group, path)?;

        if let Ok(watcher) = state.yaml_watcher.lock() {
            let _ = watcher.update_yaml_timestamp(yaml_path);
        }
    }
    Ok(())
}

pub fn update_project_in_yaml(
    group: &Group,
    _project: &Project,
    state: &crate::state::AppState,
) -> Result<(), String> {
    if !group.sync_enabled {
        return Ok(());
    }

    if let Some(ref yaml_path) = group.sync_file {
        let path = Path::new(yaml_path);
        write_yaml(group, path)?;

        if let Ok(watcher) = state.yaml_watcher.lock() {
            let _ = watcher.update_yaml_timestamp(yaml_path);
        }
    }
    Ok(())
}

pub fn remove_project_from_yaml(
    group: &Group,
    _project_id: &str,
    state: &crate::state::AppState,
) -> Result<(), String> {
    if !group.sync_enabled {
        return Ok(());
    }

    if let Some(ref yaml_path) = group.sync_file {
        let path = Path::new(yaml_path);
        write_yaml(group, path)?;

        if let Ok(watcher) = state.yaml_watcher.lock() {
            let _ = watcher.update_yaml_timestamp(yaml_path);
        }
    }
    Ok(())
}

pub fn update_group_in_yaml(group: &Group, state: &crate::state::AppState) -> Result<(), String> {
    if !group.sync_enabled {
        return Ok(());
    }

    if let Some(ref yaml_path) = group.sync_file {
        let path = Path::new(yaml_path);
        write_yaml(group, path)?;

        if let Ok(watcher) = state.yaml_watcher.lock() {
            let _ = watcher.update_yaml_timestamp(yaml_path);
        }
    }
    Ok(())
}

File: src-electron/main/services/yaml-config.ts
/**
 * YAML configuration service for reading/writing openrunner.yaml files.
 * This is the equivalent of src-tauri/src/yaml_config.rs
 */

import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'js-yaml';
import type { Group, Project, ProjectType } from '../../shared/types';

const YAML_FILENAME = 'openrunner.yaml';
const YML_FILENAME = 'openrunner.yml';

export interface YamlConfig {
  version: string;
  name: string;
  envVars?: Record<string, string>;
  projects: YamlProject[];
}

export interface YamlProject {
  name: string;
  command: string;
  type: 'service' | 'task';
  autoRestart?: boolean;
  cwd?: string;
  interactive?: boolean;
  envVars?: Record<string, string>;
  watchPatterns?: string[];
  autoStartOnLaunch?: boolean;
}

/**
 * Find an openrunner.yaml or openrunner.yml file in a directory
 */
export function findYamlFile(directory: string): string | null {
  const yamlPath = path.join(directory, YAML_FILENAME);
  if (fs.existsSync(yamlPath)) {
    return yamlPath;
  }

  const ymlPath = path.join(directory, YML_FILENAME);
  if (fs.existsSync(ymlPath)) {
    return ymlPath;
  }

  return null;
}

/**
 * Get the default YAML path for a directory (creates openrunner.yaml if neither exists)
 */
export function getYamlPath(directory: string): string {
  const yamlPath = path.join(directory, YAML_FILENAME);
  if (fs.existsSync(yamlPath)) {
    return yamlPath;
  }

  const ymlPath = path.join(directory, YML_FILENAME);
  if (fs.existsSync(ymlPath)) {
    return ymlPath;
  }

  // Default to .yaml extension
  return yamlPath;
}

/**
 * Parse a YAML config file
 */
export function parseYaml(filePath: string): YamlConfig {
  const content = fs.readFileSync(filePath, 'utf-8');
  const config = yaml.load(content) as YamlConfig;

  if (!config || typeof config !== 'object') {
    throw new Error('Invalid YAML file format');
  }

  if (!Array.isArray(config.projects)) {
    throw new Error('YAML file must contain a projects array');
  }

  return config;
}

/**
 * Write a group to a YAML file
 */
export function writeYaml(group: Group, filePath: string): void {
  const yamlProjects: YamlProject[] = group.projects.map((p) => {
    const yamlProject: YamlProject = {
      name: p.name,
      command: p.command,
      type: p.projectType,
      autoRestart: p.autoRestart,
      cwd: p.cwd ?? undefined,
      interactive: p.interactive,
    };

    // Only include envVars if non-empty
    if (Object.keys(p.envVars).length > 0) {
      yamlProject.envVars = p.envVars;
    }

    // Only include watchPatterns if defined and non-empty
    if (p.watchPatterns && p.watchPatterns.length > 0) {
      yamlProject.watchPatterns = p.watchPatterns;
    }

    // Only include autoStartOnLaunch if true
    if (p.autoStartOnLaunch) {
      yamlProject.autoStartOnLaunch = p.autoStartOnLaunch;
    }

    return yamlProject;
  });

  const yamlConfig: YamlConfig = {
    version: '1.0',
    name: group.name,
    projects: yamlProjects,
  };

  // Only include envVars if non-empty
  if (Object.keys(group.envVars).length > 0) {
    yamlConfig.envVars = group.envVars;
  }

  const yamlContent = yaml.dump(yamlConfig, {
    indent: 2,
    lineWidth: -1, // Disable line wrapping
    noRefs: true,
    quotingType: '"',
    forceQuotes: false,
  });

  fs.writeFileSync(filePath, yamlContent, 'utf-8');
}

/**
 * Convert a YAML config to a Group
 */
export function yamlToGroup(
  config: YamlConfig,
  directory: string,
  syncFile: string
): Group {
  const projects: Project[] = config.projects.map((p) =>
    yamlToProject(p, directory)
  );

  return {
    id: crypto.randomUUID(),
    name: config.name,
    directory,
    projects,
    envVars: config.envVars ?? {},
    syncFile,
    syncEnabled: true,
  };
}

/**
 * Convert a YAML project to a Project
 */
export function yamlToProject(yamlProject: YamlProject, _baseDir: string): Project {
  return {
    id: crypto.randomUUID(),
    name: yamlProject.name,
    command: yamlProject.command,
    projectType: yamlProject.type as ProjectType,
    autoRestart: yamlProject.autoRestart ?? true,
    cwd: yamlProject.cwd ?? null,
    interactive: yamlProject.interactive ?? false,
    envVars: yamlProject.envVars ?? {},
    watchPatterns: yamlProject.watchPatterns,
    autoStartOnLaunch: yamlProject.autoStartOnLaunch ?? false,
  };
}

/**
 * Update an existing group from a YAML config (preserves project IDs where names match)
 */
export function updateGroupFromYaml(
  group: Group,
  config: YamlConfig,
  baseDir: string
): Group {
  // Build a map of existing project names to IDs
  const existingProjectIds = new Map<string, string>();
  for (const project of group.projects) {
    existingProjectIds.set(project.name, project.id);
  }

  // Create new projects, preserving IDs where names match
  const projects: Project[] = config.projects.map((yp) => {
    const project = yamlToProject(yp, baseDir);
    const existingId = existingProjectIds.get(project.name);
    if (existingId) {
      project.id = existingId;
    }
    return project;
  });

  return {
    ...group,
    name: config.name,
    envVars: config.envVars ?? {},
    projects,
  };
}
